### PR TITLE
Fix BufferedStream.FlushAsync bug where the incorrect property was checked

### LINKED ISSUE #13297: Possible bug in BufferedStream.FlushAsync for read-only Streams
There is [this code](https://github.com/dotnet/corefx/blob/master/src/System.IO/src/System/IO/BufferedStream.cs#L272-L290) in the synchronous `BufferedStream.Flush`:

```cs
            if (_readPos < _readLen)
            {
                // If the underlying stream is not seekable AND we have something in the read buffer, then FlushRead would throw.
                // We can either throw away the buffer resulting in data loss (!) or ignore the Flush.
                // (We cannot throw because it would be a breaking change.) We opt into ignoring the Flush in that situation.
                if (!_stream.CanSeek)
                    return;

                FlushRead();

                // User streams may have opted to throw from Flush if CanWrite is false (although the abstract Stream does not do so).
                // However, if we do not forward the Flush to the underlying stream, we may have problems when chaining several streams.
                // Let us make a best effort attempt:
                if (_stream.CanWrite)
                    _stream.Flush();

                Debug.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
                return;
            }
```

Here is the [corresponding code](https://github.com/dotnet/corefx/blob/master/src/System.IO/src/System/IO/BufferedStream.cs#L324-L342) in the async version:

```cs
                if (_readPos < _readLen)
                {
                    // If the underlying stream is not seekable AND we have something in the read buffer, then FlushRead would throw.
                    // We can either throw away the buffer resulting in date loss (!) or ignore the Flush. (We cannot throw because it
                    // would be a breaking change.) We opt into ignoring the Flush in that situation.
                    if (!_stream.CanSeek)
                        return;

                    FlushRead();  // not async; it uses Seek, but there's no SeekAsync

                    // User streams may have opted to throw from Flush if CanWrite is false (although the abstract Stream does not do so).
                    // However, if we do not forward the Flush to the underlying stream, we may have problems when chaining several streams.
                    // Let us make a best effort attempt:
                    if (_stream.CanRead)
                        await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);

                    Debug.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
                    return;
                }
```

Note how in the first version, we check if the stream is writable before calling `Flush` while in the second `FlushAsync` is called if the stream is readable. It looks like this is a typo, and the second version should have checked `_stream.CanWrite`.

### PR DESCRIPTION
**Pull request by jamesqo**
See #13297 for more details.

Fixes #13297

@ianhays, @JeremyKuhne, @stephentoub

### CODE REVIEW

**Comment by stephentoub** [comment]
There are some failing tests related to this PR:
```
System.IO.Tests.BufferedStreamFlushTests.ShouldNotFlushUnderlyingStreamIfReadOnly [FAIL]
        System.NotSupportedException : Specified method is not supported.
        Stack Trace:
  Discovered:  System.IO.Tests
           D:\j\workspace\windows_nt_de---4526f5ff\src\Common\tests\System\IO\DelegateStream.cs(60,0): at System.IO.DelegateStream.<>c.<.ctor>b__14_9(Int64 _, SeekOrigin __)
           D:\j\workspace\windows_nt_de---4526f5ff\src\Common\tests\System\IO\DelegateStream.cs(82,0): at System.IO.DelegateStream.Seek(Int64 offset, SeekOrigin origin)
           D:\j\workspace\windows_nt_de---4526f5ff\src\Common\tests\System\IO\CallTrackingStream.cs(148,0): at System.IO.Tests.CallTrackingStream.Seek(Int64 offset, SeekOrigin origin)
              at System.IO.BufferedStream.FlushRead()
              at System.IO.BufferedStream.Flush()
           D:\j\workspace\windows_nt_de---4526f5ff\src\System.IO\tests\BufferedStream\BufferedStream.FlushTests.cs(34,0): at System.IO.Tests.BufferedStreamFlushTests.<ShouldNotFlushUnderlyingStreamIfReadOnly>d__0.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```

**Comment by ianhays** [comment]
One thing to note is that this is a divergence from the desktop behavior since we check CanRead there also.

Though it is a pretty clear bug, as even the comment mentions checking CanWrite.

**Comment by jamesqo** [comment]
@stephentoub, yes, the tests I added are failing. Strangely enough they seemed to pass on my machine, even when I undid the commit with the actual changes.

@ianhays @stephentoub 

> One thing to note is that this is a divergence from the desktop behavior since we check CanRead there also.

> Though it is a pretty clear bug, as even the comment mentions checking CanWrite.

Is it also a bug that we don't call `_stream.FlushAsync` if the stream is not seekable, but is writable? It seems the `CanSeek` check was inserted for the purpose of avoiding `FlushRead`, which is only supported for seekable streams, but then we miss out on the part that calls `_stream.FlushAsync` if `CanWrite` is true. Should the code be changed to

```cs
if (_stream.CanSeek)
{
    FlushRead();
}

if (_stream.CanWrite)
{
    await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
}
```

? Same for `Flush`.

**Comment by stephentoub** [comment]
> Is it also a bug that we don't call _stream.FlushAsync if the stream is not seekable, but is writable?

If there's something in the read buffer, then there's nothing to write.

**Comment by jamesqo** [comment]
> If there's something in the read buffer, then there's nothing to write.

@stephentoub I'm not talking about the FlushAsync call that's used to flush the write buffer, but the one that comes after `FlushRead` (the one affected by this PR). There is this snippet from the code:

```cs
                if (_readPos < _readLen)
                {
                    // If the underlying stream is not seekable AND we have something in the read buffer, then FlushRead would throw.
                    // We can either throw away the buffer resulting in date loss (!) or ignore the Flush. (We cannot throw because it
                    // would be a breaking change.) We opt into ignoring the Flush in that situation.
                    if (!_stream.CanSeek)
                        return;

                    FlushRead();  // not async; it uses Seek, but there's no SeekAsync

                    // User streams may have opted to throw from Flush if CanWrite is false (although the abstract Stream does not do so).
                    // However, if we do not forward the Flush to the underlying stream, we may have problems when chaining several streams.
                    // Let us make a best effort attempt:
                    if (_stream.CanWrite)
                        await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);

                    Debug.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
                    return;
                }

                // We had no data in the buffer, but we still need to tell the underlying stream to flush.
                if (_stream.CanWrite)
                    await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
```

So notice if there is something in the read buffer and the stream is non-seekable, we do not actually flush the stream. If there is nothing in either buffer and the stream is non-seekable, however, we do attempt to flush the stream.

Here's an example of what I'm talking about:

```cs
var s = new SomeStream { CanSeek = false, CanWrite = true, CanRead = true };
var b = new BufferedStream(s);

await b.FlushAsync(); // calls s.FlushAsync

b.ReadByte();
await b.FlushAsync(); // does not call s.FlushAsync
```

**Comment by stephentoub** [comment]
>  I'm not talking about the FlushAsync call that's used to flush the write buffer, but the one that comes after FlushRead

Ah, you're not asking about whether it's important for BufferedStream's state to be flushed to the stream it wraps in this case, but whether a call to BufferedStream.FlushAsync should do whatever it needs to and then also call _stream.FlushAsync. There are plenty of streams that don't make that guarantee, i.e. where a WrapperStream.FlushAsync call is about flushing to the wrapped stream and not also flushing the wrapped stream to whatever it sits on top of, so it's debatable whether BufferedStream should or not.  However, since BufferedStream does do so sometimes, seems reasonable to make it consistent.

**Comment by jamesqo** [comment]
Test Innerloop CentOS7.1 Debug Build and Test
Test Innerloop Windows_NT Debug Build and Test

**Comment by jamesqo** [comment]
@dotnet-bot test this please

File: src/System.IO/tests/BufferedStream/BufferedStream.FlushTests.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+namespace System.IO.Tests
+{
+    public class BufferedStreamFlushTests
+    {
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task ShouldNotFlushUnderlyingStreamIfReadOnly(bool underlyingCanSeek)
+        {
+            var underlying = new DelegateStream(
```
Comment (jamesqo) [inline src/System.IO/tests/BufferedStream/BufferedStream.FlushTests.cs]: I don't think so. I just checked and that Stream doesn't override Dispose, so I don't think there's a difference.

**Comment by jamesqo** [comment]
> Have you by any chance run these tests on the full framework? It would be nice to know what exact scenarios this would be breaking.

I tried using `/p:TestTFM=net46` and `/p:TestTFM=net463`, but the xUnit runner seems to be fizzing out with a `FileLoadException` for both of them. I do not believe this is likely to break any tests that we have in place, though.

**Comment by jamesqo** [comment]
Test Innerloop CentOS7.1 Debug Build and Test
Test Innerloop OSX Debug Build and Test 
Test Innerloop Ubuntu14.04 Debug Build and Test
Test Innerloop Windows_NT Debug Build and Test

**Comment by jamesqo** [comment]
@stephentoub @ianhays I fixed an assert related to this change that was causing debug CI to fail. This should be safe to merge now.

**Comment by ianhays** [comment]
> I do not believe this is likely to break any tests that we have in place, though.

I was more concerned that this would constitute a "breaking change" and would end up with these tests failing when run for Desktop. If that were the case, we should pre-emptively disable them for that platform and add a comment over the test explaining why it is disabled.

**Comment by jamesqo** [comment]
@ianhays, I cannot test the changes on desktop. Am running into this issue with the test runner: #13911

**Comment by stephentoub** [comment]
@ianhays, what are required next steps here?

**Pull request merged**